### DISCUSSION TITLE
Should Pinia warn when a mapState property is initially not present?

### DISCUSSION THREAD

**Discussion by rbanks54** [vuejs/pinia#641] [Help and Questions] [answered]
Quick scenario: Vue 2 with Options API, Typescript, Pinia 0.5.4

Assume I have a state defined as

```
export class MyState {
  currentId?: ReferenceId;
}
```

and define my store as
```
export const useMyStore = defineStore({
  id: 'myStore',
  state(): MyState { return new MyState() },
  actions: {
    setActiveReference(referenceId: ReferenceId){
      Vue.set(this,'currentId',referenceId);
    },
    //...
  }
}
```
and finally, I in my options API I use mapState() like so...
```
computed: {
  ...mapState(useMyStore, {
    activeReference: state => state.currentId,
//...
```

As far as I can tell, when the `mapState()` method is called, since the `currentId` property isn't present in the state object there won't be any subscription for change notifications created. `activeReference` will always be a non-reactive as a result.

Even once `setActiveReference()` is called, `activeReference` won't suddenly start being reactive because there's no way to indicate to the component that the `currentId` property now exists and change events can be subscribed to.

Since it's easy to miss explicitly setting undefined values in classes (at least, it was for me!) is it possible/worthwhile for Pinia to warn during the `mapState()` call if a keymapper function is referencing a property that doesn't actually exist?

**Comment by posva** [comment, marked as answer]
The problem here is that the return value of state should be a plain object, like `data`. Probably worth adding to the API docs. Maybe we could add a warning if the return isn't a plain object. There is already a function for it (I think it should work) in the code base !
